---
title: "chipseq_pipeline"
output: html_notebook
---

I followed the course's format, including the same sequence resource

. Config File
```{bash nextflow.config file}
# making the .config file
nano nextflow.config

# Paste in the following code into the nextflow.config file

# I have to add thie line here, otherwise there's error because the new nextflow version does not support dsl1
# honestly i don't actually understand, but the error disappeared once i added this line
nextflow.enable.dsl=2
process {
  executor='slurm'
  queue='short'
  memory='32 GB'
  maxForks=10
}
```


. Design file 

These are the file paths for all the fastq files. I followed the course's format and downloaded the same files.
Got all the files from here: https://www.encodeproject.org/
```{bash design file}
nano design.csv

# copy the following to design.csv
group,replicate,fastq_1,fastq_2,antibody,control
POLR2A,3,/scratch1/difeizhu/practice/fastq/ENCFF000XWJ.fastq.gz,,POLR2A,ENCSR000EEN
POLR2A,2,/scratch1/difeizhu/practice/fastq/ENCFF000PNT.fastq.gz,,POLR2A,ENCSR000BLH
POLR2A,4,/scratch1/difeizhu/practice/fastq/ENCFF000XWR.fastq.gz,,POLR2A,ENCSR000EEN
POLR2A,1,/scratch1/difeizhu/practice/fastq/ENCFF000PNM.fastq.gz,,POLR2A,ENCSR000BLH
ENCSR000EEN,1,/scratch1/difeizhu/practice/fastq/ENCFF000XTF.fastq.gz,,,
ENCSR000BLH,1,/scratch1/difeizhu/practice/fastq/ENCFF000POQ.fastq.gz,,,
```



. Shell script 

This is my version:
Things I added/changed: 
#SBATCH --account=aealmada_561
#SBATCH --partition=main
#SBATCH --mail-user=difeizhu@usc.edu
got fasta file from here: https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/
got gtf file from here: https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/
got blacklist file from here: https://www.dropbox.com/s/v6uo554wgq4arb5/hg38-blacklist.v2.bed?dl=0
```{bash}
#!/bin/bash
#SBATCH --account=aealmada_561
#SBATCH --partition=main
#SBATCH --job-name=PolII_test
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=difeizhu@usc.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=6gb
#SBATCH --time=14:00:00
#SBATCH --output=nextflow.out
#SBATCH --error=nextflow.err
pwd; hostname; date
echo "Lets go"
nextflow run nf-core/chipseq -r 1.2.2 \
-profile singularity \
--single_end \
--input design.csv \
--fasta /scratch1/difeizhu/practice/gencode/GRCh38.p13.genome.fa \
--gtf /scratch1/difeizhu/practice/gencode/gencode.v32.annotation.gtf \
--macs_gsize 3.2e9 \
--blacklist /scratch1/difeizhu/practice/gencode/hg38-blacklist.v2.bed \
--email difeizhu@usc.edu \
-resume \
-c nextflow.config
date

```


This is the course's version
```{r}
#!/bin/bash
#SBATCH -p short
#SBATCH --job-name=Hepg2_Pol_test
#SBATCH --mail-type=END,FAIL
#SBATCH --mail-user=your_email
#SBATCH --output=nextflow.out
#SBATCH --error=nextflow.err
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=6gb
#SBATCH --time=14:00:00
pwd; hostname; date
echo "Lets do chipseq"
module load singularity/3.1.1
nextflow run nf-core/chipseq -r 1.2.1 \
-profile singularity \
--single_end \
--input design.csv \
--fasta /scratch/Shares/rinnclass/CLASS_2023/data/data/genomes/human/gencode/v32/GRCh38.p13.genome.fa \
--gtf /scratch/Shares/rinnclass/CLASS_2023/data/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf \
--macs_gsize 3.2e9 \
--blacklist /scratch/Shares/rinnclass/CLASS_2023/data/data/genomes/hg38-blacklist.v2.bed \
--email your email \
-resume \
-c nextflow.config
date
```



submit the job

```{r}
sbatch run.sh
```




***************
Troubleshooting

Will generate following files/folders:
nextflow.err   (for error messages)
nextflow.out  (for error messages and tracking)
results 
work


Error messages: 
ERROR ~ No such variable: ch_design_reads_csv

 -- Check script '/home1/difeizhu/.nextflow/assets/nf-core/chipseq/main.nf' at line: 340 or see '.nextflow.log' file for more details



I found the main.nf file and checked it. It's inside the home1 directory.
```{r}
cd /home1/difeizhu/.nextflow/assets/nf-core/chipseq
ls
cat main.nf
```

also found the .nextflow.log file. It's inside the chipseq pipeline working directory.
```{r}
cd /scratch1/difeizhu/practice
cat .nextflow.log
```



